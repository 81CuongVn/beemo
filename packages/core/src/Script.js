/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Module } from 'boost';

import type { Tool, Reporter } from 'boost';
import type Driver from './Driver';

export default class Script<To: Object = {}> extends Module<To> {
  /**
   * Run the script with the options object and Beemo tool instance.
   */
  run(options: Object, tool: Tool<Driver, Reporter<Object>>): Promise<*> {
    throw new Error(`${this.constructor.name} must define a run() method.`);
  }

  /**
   * Pass a configuration object to parse args with.
   */
  setup(): Object {
    return {};
  }
}
