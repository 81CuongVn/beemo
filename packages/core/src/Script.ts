/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Module, Tool } from '@boost/core';
import { Options, Arguments } from 'yargs';

export default class Script<Opts = {}> extends Module<Opts> {
  /**
   * Define a configuration object to parse args with.
   */
  parse(): { [option: string]: Options } {
    return {};
  }

  /**
   * Run the script with the options object and Beemo tool instance.
   */
  async run(args: Arguments, tool: Tool): Promise<any> {
    throw new Error(`${this.constructor.name} must define an async run() method.`);
  }
}
