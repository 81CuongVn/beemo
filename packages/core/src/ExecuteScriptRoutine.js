/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import path from 'path';
import { ModuleLoader, Routine } from 'boost';
import parseArgs from 'yargs-parser';
import Script from './Script';

import type { BeemoConfig, Execution, ScriptContext } from './types';

export default class ExecuteScriptRoutine extends Routine<BeemoConfig, ScriptContext> {
  execute(scriptName: string): Promise<string[]> {
    this.task('Loading script', this.loadScript);
    this.task('Running script', this.runScript);

    return this.serializeTasks(scriptName);
  }

  /**
   * Attempt to load a script from the configuration module.
   */
  loadScript(scriptName: string): Promise<Script> {
    const filePath = path.join(this.context.moduleRoot, 'scripts', `${scriptName}.js`);
    const loader = new ModuleLoader(this.tool, 'script', Script);

    this.tool.debug('Loading script');

    return new Promise(resolve => {
      const script = loader.importModule(filePath);

      // Is not set by Boost, so set it here
      script.name = scriptName;

      this.context.script = script;
      this.context.scriptName = scriptName;
      this.context.scriptPath = filePath;

      resolve(script);
    });
  }

  /**
   * Run the script while also parsing arguments to use as options.
   */
  runScript(script: Script): Promise<Execution> {
    const { args, yargs } = this.context;

    this.tool.debug(`Executing script with args "${args.join(' ')}"`);

    this.tool.emit('execute-script', [script, args, yargs]);

    const options = parseArgs(args, script.parse());

    return Promise.resolve(script.run(options, this.tool))
      .then(response => {
        this.tool.emit('successful-script', [script, response]);

        return response;
      })
      .catch(error => {
        this.tool.emit('failed-script', [script, error]);

        throw error;
      });
  }
}
