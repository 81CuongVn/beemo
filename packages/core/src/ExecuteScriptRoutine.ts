/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import path from 'path';
import { ModuleLoader, Routine } from '@boost/core';
import parseArgs from 'yargs-parser';
import Script from './Script';
import ScriptContext from './contexts/ScriptContext';
import { BeemoTool, Execution } from './types';

export default class ExecuteScriptRoutine extends Routine<ScriptContext, BeemoTool> {
  bootstrap() {
    this.task(this.tool.msg('app:scriptLoad'), this.loadScript);
    this.task(this.tool.msg('app:scriptRun'), this.runScript);
  }

  execute(): Promise<Execution> {
    return this.serializeTasks();
  }

  /**
   * Attempt to load a script from the configuration module.
   */
  async loadScript(context: ScriptContext): Promise<Script> {
    const filePath = path.join(context.moduleRoot, 'scripts', `${context.scriptName}.js`);
    const loader = new ModuleLoader(this.tool, 'script', Script);
    let script: Script;
    let loadError: Error;

    this.debug('Loading script from configuration module');

    try {
      script = loader.importModule(filePath, [
        context.scriptName,
        this.tool.msg('app:scriptRunNamed', { name: context.scriptName }),
      ]);
    } catch (error) {
      loadError = error;
      this.debug('Failed to load from configuration module: %s', error.message);
    }

    // Pass context and tool to script
    script.configure(this);

    // Set script into context
    context.setScript(script, filePath);

    this.tool.emit(`${context.eventName}.load-script`, [context, script]);

    return script;
  }

  /**
   * Run the script while also parsing arguments to use as options.
   */
  async runScript(context: ScriptContext, script: Script): Promise<Execution> {
    const { argv } = this.context;

    this.debug('Executing script with args "%s"', argv.join(' '));

    this.tool.emit(`${context.eventName}.before-execute`, [context, argv, script]);

    const args = parseArgs(argv, script.args());
    let result = null;

    try {
      result = await script.execute(context, args);

      this.tool.emit(`${context.eventName}.after-execute`, [context, result, script]);
    } catch (error) {
      this.tool.emit(`${context.eventName}.failed-execute`, [context, error, script]);

      throw error;
    }

    return result;
  }
}
