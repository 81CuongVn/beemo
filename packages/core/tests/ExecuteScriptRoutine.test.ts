import ModuleLoader from '@boost/core/lib/ModuleLoader';
import ExecuteScriptRoutine from '../src/ExecuteScriptRoutine';
import Script from '../src/Script';
import {
  prependRoot,
  createScriptContext,
  createTestDebugger,
  createTestTool,
} from '../../../tests/helpers';

jest.mock('@boost/core/lib/ModuleLoader', () =>
  jest.fn(() => ({
    importModule: jest.fn(() => ({
      parse: () => ({}),
      run: () => Promise.resolve(123),
    })),
  })),
);

describe('ExecuteScriptRoutine', () => {
  let routine: ExecuteScriptRoutine;

  beforeEach(() => {
    routine = new ExecuteScriptRoutine('script', 'Executing script');
    routine.context = createScriptContext();
    routine.tool = createTestTool();
    routine.debug = createTestDebugger();

    // @ts-ignore
    ModuleLoader.mockClear();
  });

  describe('execute()', () => {
    it('passes script name to tasks', async () => {
      routine.serializeTasks = jest.fn();

      await routine.execute(routine.context, 'foo');

      expect(routine.serializeTasks).toHaveBeenCalledWith('foo');
    });

    it('executes pipeline in order', async () => {
      const loadSpy = jest.spyOn(routine, 'loadScript');
      const runSpy = jest.spyOn(routine, 'runScript');

      const response = await routine.execute(routine.context, 'foo-bar');

      expect(loadSpy).toHaveBeenCalledWith(routine.context, 'foo-bar', expect.anything());
      expect(runSpy).toHaveBeenCalledWith(
        routine.context,
        expect.objectContaining({
          name: 'foo-bar',
        }),
        expect.anything(),
      );
      expect(response).toBe(123);
    });
  });

  describe('loadScript()', () => {
    it('loads the script using ModuleLoader', async () => {
      await routine.loadScript(routine.context, 'foo-bar');

      expect(ModuleLoader).toHaveBeenCalledWith(routine.tool, 'script', Script);
    });

    it('sets values to context', async () => {
      const script = await routine.loadScript(routine.context, 'foo-bar');

      expect(script.name).toBe('foo-bar');
      expect(routine.context).toEqual(
        expect.objectContaining({
          scriptName: 'foo-bar',
          scriptPath: prependRoot('scripts/foo-bar.js'),
        }),
      );
    });

    it('triggers `load-script` event', async () => {
      const spy = jest.spyOn(routine.tool, 'emit');

      const script = await routine.loadScript(routine.context, 'foo-bar');

      expect(spy).toHaveBeenCalledWith('foo-bar.load-script', [script]);
    });
  });

  describe('runScript()', () => {
    it('calls the scripts parse() and run()', async () => {
      const script = new Script();
      script.parse = jest.fn(() => ({
        boolean: ['foo'],
        default: {
          foo: false,
        },
      }));
      script.run = jest.fn();

      await routine.runScript(routine.context, script);

      expect(script.parse).toHaveBeenCalled();
      expect(script.run).toHaveBeenCalledWith(
        expect.objectContaining({
          _: ['bar', 'baz'],
          a: true,
          foo: true,
        }),
        routine.tool,
      );
    });

    it('triggers `before-execute` event', async () => {
      class MockScript extends Script {
        name = 'before';

        run() {
          return Promise.resolve();
        }
      }

      const spy = jest.spyOn(routine.tool, 'emit');
      const script = new MockScript();

      await routine.runScript(routine.context, script);

      expect(spy).toHaveBeenCalledWith('before.before-execute', [
        script,
        routine.context.argv,
        routine.context,
      ]);
    });

    it('triggers `after-execute` event on success', async () => {
      class SuccessScript extends Script {
        name = 'after';

        run() {
          return Promise.resolve(123);
        }
      }

      const spy = jest.spyOn(routine.tool, 'emit');
      const script = new SuccessScript();

      await routine.runScript(routine.context, script);

      expect(spy).toHaveBeenCalledWith('after.after-execute', [script, 123]);
    });

    it('triggers `failed-execute` event on failure', async () => {
      class FailureScript extends Script {
        name = 'fail';

        run() {
          return Promise.reject(new Error('Oops'));
        }
      }

      const spy = jest.spyOn(routine.tool, 'emit');
      const script = new FailureScript();

      try {
        await routine.runScript(routine.context, script);
      } catch (error) {
        expect(spy).toHaveBeenCalledWith('fail.failed-execute', [script, error]);
      }
    });
  });
});
