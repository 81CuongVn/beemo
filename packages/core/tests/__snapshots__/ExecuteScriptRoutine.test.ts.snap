// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExecuteScriptRoutine loadScript() errors when neither file path or module can be loaded 1`] = `
"Failed to load script, the following errors occurred:
Script \\"Missing.js\\" missing!
Script \\"missing\\" missing!"
`;
